### DISCUSSION TITLE
Custom filter for processing Github/Obsidian style callouts stopped working - any tips?

### DISCUSSION THREAD

**Discussion by danieltomasz** [quarto-dev/quarto-cli#6550] [Q&A] [answered]
### Description

In order to process a obsidian/new github style callouts 
```
> [!NOTE]
> Highlights information that users should take into account, even when skimming.
```

I added a custom lua filter to my `_quarto.yml`
with
```
filters:
- obsidian-callouts.lua
```

```
local stringify = (require "pandoc.utils").stringify

function BlockQuote (el)
    start = el.content[1]
    if (start.t == "Para" and start.content[1].t == "Str" and
        start.content[1].text:match("^%[!%w+%][-+]?$")) then
        _, _, ctype = start.content[1].text:find("%[!(%w+)%]")
        el.content:remove(1)
        start.content:remove(1)
        local class = "callout-" .. ctype:lower()
        div = pandoc.Div(el.content, {class = class})
        div.attributes["data-callout"] = ctype:lower()
        div.attributes["title"] = stringify(start.content):gsub("^ ", "")
        return div
    else
        return el
    end
end

-- source https://forum.obsidian.md/t/rendering-callouts-similarly-in-pandoc/40020/6
```
the filter worked a few months ago with 1.3 and previous pandoc, but now the filter produce empty output (no text from callout)
Anyone has any idea how to fix it?

**Comment by cscheid** [comment]
Can you provide a full example?

**Comment by danieltomasz** [comment]
I have now a problem with fully reproducing it. The behaviour I am expecting was working in May (colorful callouts), I even wrote a post about it
https://danielborek.me/2023/obsidian-quarto-callouts/

To partially replicate here are file
```test.qmd
---
title: "Untitled"
format: html
filters:
  - callouts.lua
---

# Test filter

> [!note] My note
>
> Note content
```
And  again the callout included
```lua
local stringify = (require "pandoc.utils").stringify

function BlockQuote (el)
    start = el.content[1]
    if (start.t == "Para" and start.content[1].t == "Str" and
        start.content[1].text:match("^%[!%w+%][-+]?$")) then
        _, _, ctype = start.content[1].text:find("%[!(%w+)%]")
        el.content:remove(1)
        start.content:remove(1)
        div = pandoc.Div(el.content, {class = "callout"})
        div.attributes["data-callout"] = ctype:lower()
        div.attributes["title"] = stringify(start.content):gsub("^ ", "")
        return div
    else
        return el
    end
end
```
The closest what I want to achieve I am getting with Quarto 1.2.475 and v1.1.189 (but previously I was able to get colour and title as well whem those are one of 5 defined  in quarto types)

<img width="792" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/7980381/a6ff4ed1-63b8-47ec-be04-8be0c0a8d528">

with the corresponding latex (I copying only the body, what is important 1.3 and 1.4 dont' produce tcolorbox)
```
\begin{document}
\maketitle
\ifdefined\Shaded\renewenvironment{Shaded}{\begin{tcolorbox}[interior hidden, borderline west={3pt}{0pt}{shadecolor}, frame hidden, enhanced, boxrule=0pt, sharp corners, breakable]}{\end{tcolorbox}}\fi

\hypertarget{test-filter}{%
\section{Test filter}\label{test-filter}}

\begin{tcolorbox}[enhanced jigsaw, opacityback=0, colback=white, rightrule=.15mm, arc=.35mm, colframe=quarto-callout-color-frame, toprule=.15mm, breakable, left=2mm, bottomrule=.15mm, leftrule=.75mm]

Note content

\end{tcolorbox}



\end{document}
```

The latest 1.3 and  1.4 releases render just text without a callout around it 
```
\begin{document}
\maketitle
\hypertarget{test-filter}{%
\section{Test filter}\label{test-filter}}

Note content



\end{document}
```
Both rendering from qmd and ipynb gave me the same results,

**Reply by cscheid** [reply, marked as answer]
Ok, I understand now what you're asking. `Callout` is now a node type in the Quarto AST, and you just need to use `quarto.Callout` explicitly:

https://quarto.org/docs/prerelease/1.3/custom-ast-nodes/callout.html

```lua
function BlockQuote (el)
    start = el.content[1]
    if (start.t == "Para" and start.content[1].t == "Str" and
        start.content[1].text:match("^%[!%w+%][-+]?$")) then
        _, _, ctype = start.content[1].text:find("%[!(%w+)%]")
        el.content:remove(1)
        return quarto.Callout({
            type = ctype:lower(),
            title = stringify(start.content):gsub("^ ", "")
            content = el.content
        })
    else
        return el
    end
```

I haven't tested this, so it might need tweaking. But the general idea should work.

**Reply by danieltomasz** [reply]
Thanks, this is really informative.

I almost got it working :) 
I have the filter as below

```lua
local stringify = (require "pandoc.utils").stringify

function BlockQuote(el)

    local start = el.content[1]

    if (start.t == "Para" and start.content[1].t == "Str" and
        start.content[1].text:match("^%[!%w+%][-+]?$")) then
        local firstline = stringify(start.content[1])
        local  _, _, ctype = firstline:find("%[!(%w+)%]")
        local titlevar =stringify(start.content):match("^%[!%w+%](.-)$")
        el.content:remove(1)
        -- Create Quarto callout
        return quarto.Callout({
          type = ctype:lower(),
          title = titlevar,
          content = el.content
        })
    else
        return el
    end
end
```


with it,  how to make the content from 2 lines
```
> Line 1
> Line 2
```

To appear in the final document  as separate  lines 
```
Line 1
Line 2
```
not
``` 
Line 1 Line 2
```
?

**Reply by danieltomasz** [reply]
I am don't know lua, how I can declare quarto dependency at the beginning to make it "proper' ?:P

**Reply by cscheid** [reply]
I assume this will work:

```
> Line 1
>
> Line 2
```

https://pandoc.org/lua-filters.html#type-blockquote shows that a BlockQuote holds a list of Block elements, so this should work (I again haven't tested).

If you're going down this rabbit hole, Pandoc's [Lua filters](https://pandoc.org/lua-filters.html) page is indispensable. You should also get in the habit of reading the AST produced by Pandoc:

```lua
-- This is Lua code

-- this works, but might be a little hard to read
print(blockquote)
-- this produces indented output, but it only works for entire Pandoc documents,
-- so you'll need to be a bit more careful with it.
print(pandoc.write(pandoc.Pandoc({blockquote}), "native"))
```

**Reply by cscheid** [reply]
> I am don't know lua, how I can declare quarto dependency at the beginning to make it "proper' ?:P

Sorry, I truly don't understand your question. What do you think is not proper about your filter declaration?

**Reply by danieltomasz** [reply]
>Sorry, I truly don't understand your question. What do you think is not proper about your filter declaration?

Everything is alright, it was just some linter in my editor complaining 
```
Undefined global `quarto`.Lua Diagnostics.(undefined-global)
```
And I was indeed not clear what I meant, I just want to declare that this filter needs `quarto` and I am quite new to lua and I didnt know how exactly t I should refer to module/submodule, but this is just a small cosmetic detail, as the current script works!


The above way to separate lines works as well :) 
```
> Line 1
>
> Line 2
```

Thank you for all the  tips and advices!

**Reply by cscheid** [reply]
> And I was indeed not clear what I meant, I just want to declare that this filter needs `quarto` and I am quite new to lua and I didnt know how exactly t I should refer to module/submodule, but this is just a small cosmetic detail, as the current script works!

Ah, I got it. No worries. Currently, there's no great way to do what you want that is independent of your editor.

However, if you use VS Code and our quarto extension, I believe we will add the typing information about the `quarto` object to Lua's LSP, so that .lua files in quarto projects behave well.

**Comment by RoyiAvital** [comment]
@danieltomasz , Is this available as an official extension?
It would be great!

**Reply by danieltomasz** [reply]
@RoyiAvital  I might try to work on it,  in meantime you might just use downloaded filter https://gist.github.com/danieltomasz/31d298aca2969adaf60d8841b68005e2

**Reply by f0nzie** [reply]
> @RoyiAvital I might try to work on it, in meantime you might just use downloaded filter https://gist.github.com/danieltomasz/31d298aca2969adaf60d8841b68005e2

The script has two extras `end` at the bottom. After removing them, it works perfect!
